heir_py/mlir_emitter.py

```diff
@@ -25,38 +116,60 @@ def emit(self):
     # TODO(#1162): get inferred or explicit return types
     return_types_str = "i64"
 
-    body = self.emit_body()
+    body = self.emit_blocks()
 
     mlir_func = f"""func.func @{func_name}({args_str}) -> ({return_types_str}) {{
 {textwrap.indent(body, '  ')}
 }}
 """
     return mlir_func
 
-  def emit_body(self):
+  def emit_blocks(self):
     blocks = self.ssa_ir.blocks
-    str_blocks = []
-    first = True
 
+    # collect loops and block header needs
+    self.omit_block_header[sorted(blocks.items())[0][0]] = True
     for block_id, block in sorted(blocks.items()):
-      instructions = []
-      for instr in block.body:
-        result = self.emit_instruction(instr)
-        if result:
-          instructions.append(result)
+      for i in range(len(block.body)):
+        # Detect a range call
+        instr = block.body[i]
+        if is_range_call(instr, self.ssa_ir):
+          loop = build_loop_from_call(i, block.body, blocks, self.ssa_ir)
+          self.loops[instr.target] = loop
+          self.omit_block_header[loop.header.next_id] = True
 
-      if first:
-        first = False
+    # print blocks
+    str_blocks = []
+    for block_id, block in sorted(blocks.items()):
+      if block_id in self.printed_blocks:
```

Comment:
Could you add a comment here explaining why the program might revisit a block multiple times? (b/c emit_loop prints multiple blocks)

Another option is to start by putting the blocks in a `deque`, change the for loop to a `while blocks_deque`, and then pass the `deque` to `emit_loop` and `emit_block` so it can `popleft` from the deque as it prints stuff out.

The reason I think this might be cleaner is because it must be true that the loop blocks are contiguous until the loop ends. The invariant that would be preserved is: the `blocks_deque` contains as its first element the next block that hasn't yet been emitted.